sips/fork_support.md

```diff
@@ -20,56 +20,260 @@ Client implementation should calculate when a new duty starts to decide if it st
+		return mainnetForks()
+	default:
+		return []*ForkData{
+			{
+				Epoch:  0,
+				Domain: DomainType{0x0, 0x0, byte(networkID), 0x0},
+			},
+		}
+	}
 }
 
-func (domainType DomainType) GetForksData() []*ForkData {
-    switch domainType.GetChain() {
-    case GenesisChain:
-        return genesisForks()
-    default:
-        return []*ForkData{}
-    }
+// mainnetForks returns all forks for the mainnet chain
+func mainnetForks() []*ForkData {
+	return []*ForkData{
+		{
+			Epoch:  0,
+			Domain: GenesisMainnet,
+		},
+	}
 }
 
-func genesisForks() []*ForkData {
-    return []*ForkData{
-        {
-            Epoch:  0,
-            Domain: GenesisMainnet,
-        },
-    }
+func (networkID NetworkID) DefaultFork() *ForkData {
+	return networkID.GetForksData()[0]
 }
 
+// GetCurrentFork returns the ForkData with highest Epoch smaller or equal to "epoch"
+func (networkID NetworkID) ForkAtEpoch(epoch phase0.Epoch) (*ForkData, error) {
+	// Get list of forks
+	forks := networkID.GetForksData()
+
+	// If empty, raise error
+	if len(forks) == 0 {
+		return nil, errors.New("GetCurrentFork: fork list by GetForksData is empty.")
+	}
+
+	var current_fork *ForkData
+	for _, fork := range forks {
+		if fork.Epoch <= epoch {
+			current_fork = fork
+		}
+	}
+	return current_fork, nil
+}
+
+func (f ForkData) GetRoot() ([]byte, error) {
+	byts, err := json.Marshal(f)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal ForkData")
+	}
+	ret := sha256.Sum256(byts)
+	return ret[:], nil
+}
+```
+
+### Structures access to types
+
+- `Share`: It's more correct for the structure to hold the information on the network that it's in, instead of the network and fork number.
+    Thus, _Share_ will contain a `NetworkID` field instead of a `DomainType` field.
+- `Config`: Similarly, _Config_ should have a `NetworkID` instead of `DomainType`.
+- `Controller`: It may drop its `DomainType` field since it has access to `config` and `Share`.
+
+Regarding the indirect usage of Identifiers:
+- `Instance` must keep its identifier to send properly formed messages.
+- `Controller` may keep its identifier, but its _StartNewInstance_ method should receive as an argument the new identifier for the instance, as well as the new _DomainType_ to update its _config_. The controller, then, updates its identifier.
+
+```go
+func (c *Controller) StartNewInstance(height Height, value []byte, identifier []byte, domainType types.DomainType) error { // <-- identifier as argument
+	if err := c.GetConfig().GetValueCheckF()(value); err != nil {
+		return errors.Wrap(err, "value invalid")
+	}
+
+	// can't use <= because of height == 0 case
+	if height < c.Height {
+		return errors.New("attempting to start an instance with a past height")
+	}
+
+	// covers height == 0 case
+	if c.StoredInstances.FindInstance(height) != nil {
+		return errors.New("instance already running")
+	}
+
+    c.config.SetSignatureDomainType(domainType) // <-- updates its config's DomainType
```

Comment:
is that a must? config is an interface provided on NewController, can't we just change it on implementation and it will propagate to controller?

Reply:
The controller is the first structure in the hierarchy to contain the config. Though config is passed as an argument, the runner doesn't have access to it and thus the Controller needs to be the one to update it. What do you think?

Reply:
LGTM

It should be updated everytime a new instance starts because a fork may get activated.